File: models/email.py
"""邮件模型"""
from pydantic import BaseModel, field_validator
from typing import List, Optional, Dict, Any, Literal
from datetime import datetime


class EmailAttachment(BaseModel):
    """邮件附件模型"""
    filename: str
    file_path: str
    file_size: int
    content_type: str


class EmailCreate(BaseModel):
    """创建邮件请求模型"""
    to_user_ids: Optional[List[str]] = None  # 点对点发送的收件人ID列表
    to_user_type: Optional[Literal["all", "students", "teachers"]] = None  # 批量发送的用户类型
    to_class_names: Optional[List[str]] = None  # 特定班级（普通管理员）
    to_grades: Optional[List[str]] = None  # 特定年级（普通管理员）
    subject: str
    content: str
    markdown_content: Optional[str] = None
    priority: Literal["low", "normal", "high", "urgent"] = "normal"
    is_relationship_required: bool = True  # 是否需要建立关系（用户间发送时）
    
    @field_validator("to_user_ids", "to_user_type")
    def validate_recipients(cls, v, info):
        """验证收件人设置：要么指定用户ID列表，要么指定用户类型"""
        data = info.data
        if not v and not data.get("to_user_ids") and not data.get("to_user_type"):
            raise ValueError("必须指定收件人（to_user_ids 或 to_user_type）")
        return v


class EmailDraftCreate(BaseModel):
    """创建草稿请求模型"""
    to_user_ids: Optional[List[str]] = None
    to_user_type: Optional[Literal["all", "students", "teachers"]] = None
    to_class_names: Optional[List[str]] = None
    to_grades: Optional[List[str]] = None
    subject: str = ""
    content: str = ""
    markdown_content: Optional[str] = None
    priority: Literal["low", "normal", "high", "urgent"] = "normal"


class EmailResponse(BaseModel):
    """邮件响应模型"""
    id: str
    from_user_id: str
    from_username: str
    to_user_ids: List[str]
    to_user_type: Optional[str] = None
    to_class_names: Optional[List[str]] = None
    to_grades: Optional[List[str]] = None
    subject: str
    content: str
    markdown_content: Optional[str] = None
    attachments: List[EmailAttachment] = []
    priority: str
    status: Literal["draft", "sent", "deleted"]
    is_relationship_required: bool
    relationship_invitation_id: Optional[str] = None  # 关联的关系邀请ID（如果是关系邀请邮件）
    created_at: str
    sent_at: Optional[str] = None
    updated_at: str


class EmailListItem(BaseModel):
    """邮件列表项模型"""
    id: str
    from_user_id: str
    from_username: str
    subject: str
    content_preview: str  # 内容预览（前100字符）
    content: Optional[str] = None  # 完整内容（用于对话窗口）
    priority: str
    is_read: bool
    created_at: str
    sent_at: Optional[str] = None
    folder: Literal["inbox", "sent", "draft", "trash"]


class EmailListResponse(BaseModel):
    """邮件列表响应模型"""
    emails: List[EmailListItem]
    total: int
    unread_count: int
    page: int = 1
    page_size: int = 20


class BatchEmailCreate(BaseModel):
    """批量发送邮件请求模型（管理员）"""
    to_user_type: Literal["all", "students", "teachers"]
    to_class_names: Optional[List[str]] = None
    to_grades: Optional[List[str]] = None
    subject: str
    content: str
    markdown_content: Optional[str] = None
    priority: Literal["low", "normal", "high", "urgent"] = "normal"

File: web/types/email.ts
export type EmailPriority = "low" | "normal" | "high" | "urgent";
export type EmailStatus = "draft" | "sent" | "deleted";
export type EmailFolder = "inbox" | "sent" | "draft" | "trash";
export type EmailTargetAudience = "all" | "students" | "teachers";

export interface EmailAttachment {
  filename: string;
  file_path: string;
  file_size: number;
  content_type: string;
}

export interface Email {
  id: string;
  from_user_id: string;
  from_username: string;
  to_user_ids: string[];
  to_user_type?: EmailTargetAudience;
  to_class_names?: string[];
  to_grades?: string[];
  subject: string;
  content: string;
  markdown_content?: string;
  attachments: EmailAttachment[];
  priority: EmailPriority;
  status: EmailStatus;
  is_relationship_required: boolean;
  relationship_invitation_id?: string;  // 关联的关系邀请ID（如果是关系邀请邮件）
  created_at: string;
  sent_at?: string;
  updated_at: string;
}

export interface EmailListItem {
  id: string;
  from_user_id: string;
  from_username: string;
  subject: string;
  content_preview: string;
  content?: string;  // 完整内容（用于对话视图）
  priority: EmailPriority;
  is_read: boolean;
  created_at: string;
  sent_at?: string;
  folder: EmailFolder;
}

export interface EmailListResponse {
  emails: EmailListItem[];
  total: number;
  unread_count: number;
  page: number;
  page_size: number;
}

export interface EmailCreateRequest {
  to_user_ids?: string[];
  to_user_type?: EmailTargetAudience;
  to_class_names?: string[];
  to_grades?: string[];
  subject: string;
  content: string;
  markdown_content?: string;
  priority?: EmailPriority;
  is_relationship_required?: boolean;
}

export interface EmailDraftCreateRequest {
  to_user_ids?: string[];
  to_user_type?: EmailTargetAudience;
  to_class_names?: string[];
  to_grades?: string[];
  subject: string;
  content: string;
  markdown_content?: string;
  priority?: EmailPriority;
}

export interface BatchEmailCreateRequest {
  to_user_type: EmailTargetAudience;
  to_class_names?: string[];
  to_grades?: string[];
  subject: string;
  content: string;
  markdown_content?: string;
  priority?: EmailPriority;
}

